packages/sanity/src/core/store/_legacy/document/getPairListener.ts

```diff
@@ -79,18 +87,55 @@ export function getPairListener(
           effectFormat: 'mendoza',
           tag: options.tag || 'document.pair-listener',
         },
-      ) as Observable<WelcomeEvent | MutationEvent | ReconnectEvent>,
-  ).pipe(
-    concatMap((event) =>
-      event.type === 'welcome'
+      )
+      .pipe(
+        shareReplayLatest({
+          predicate: (event) => event.type === 'welcome' || event.type === 'reconnect',
+          resetOnRefCountZero: () => timer(LISTENER_RESET_DELAY),
+        }),
+      ),
+  ) as Observable<WelcomeEvent | MutationEvent | ReconnectEvent>
+
+  return defer(() => sharedEvents).pipe(
+    concatMap((event) => {
+      return event.type === 'welcome'
         ? fetchInitialDocumentSnapshots().pipe(
-            concatMap((snapshots) => [
-              createSnapshotEvent(draftId, snapshots.draft),
-              createSnapshotEvent(publishedId, snapshots.published),
-            ]),
+            concatMap((snapshots) => {
+              return merge(
+                concat(
+                  of(createSnapshotEvent(draftId, snapshots.draft)),
+                  sharedEvents.pipe(
+                    filter((event) => event.type !== 'welcome'),
+                    filter(
+                      (sharedListenerEvent) =>
+                        sharedListenerEvent.type !== 'mutation' ||
+                        sharedListenerEvent.documentId === draftId,
+                    ),
```

Comment:
can/should these filter statements be combined? it took me a second to understand what's going on so may be nice to refactor for readability
```suggestion
                    filter((e) => {
                      switch (e.type) {
                        case 'welcome': return false
                        case 'reconnect': return true
                        case 'mutation': return e.documentId === draftId
                        default: return false
                      }
                    })
```

Reply:
I agree it's not the most readable. Your suggestion is strictly not equal since it will filter out any non-mutation event, while the original code will include any non-mutation event. Let's revisit this later, ok? A comment might help here as well.